aten/src/ATen/core/ivalue_inl.h

```diff
@@ -411,6 +411,11 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
     return fut;
   }
 
+  // Since this file cannot import CUDA depedency, use void* to represent CUDA
+  // stream, and will cast it into the correct stream type whenever use it.
+  virtual void setRecordStreamCallback(
```

Comment:
You can change the signature of this method to the following to address the CUDA stream issue:

```
template< typename T>
virtual void setRecordStreamCallback(T record_stream_cb) {}
```

This is also slightly more efficient than std::function and I think you can pass in a lambda here for `record_stream_cb` which accepts a CUDA stream as input.

Reply:
Please note that setRecordStreamCallback is defined inside ivalue::Future. Do we want to make ivalue::Future a template class? If so, how shall we instantiate T for other backends like Gloo and MPI?

Another concern: I am a fan of template class, but making this core class as a template class for only such a minor method may not justify the loss of debug-ability to some extent, due to the notorious error messages. As I can see, this class seems to be updated quite a few times recently.

Reply:
Alternatively, what about replacing void* and reinterpret_cast with std::any and any_cast? This is a C++17 feature, and I am glad that our codebase can support C++17.

Reply:
Per offline discussion, used c10::Stream to construct CUDAStream on the fly and avoid the usage of reinterpret_cast.